### PR TITLE
fix: widen the package manager fallback guard in the CLI

### PR DESCRIPTION
**Pull request by vlaux**
## What's the purpose of this pull request?

Backport of the `3.x` half of FAS-1199. The `dev` counterpart is #3422.

A store build failed with `/bin/sh: syntax error: unexpected "("` and exit code 2, with nothing in the log tying the error to its cause. The store runs on `3.x`, and no pipeline-side mitigation is planned, so this backport is the only thing that reaches stores on this line.

`getPreferredPackageManager()` returns the stdout of `na ?` and callers interpolate it into shell commands. The existing guard only covers **empty** output:

```ts
if (agent === '') return 'yarn'
```

But `ni` does not fail when the detected package manager is missing from `PATH`. It renders a confirm prompt to stdout, and that text contains `(`:

```
sh -c "Would you like to globally install pnpm (https://…)? › (y/N) run build"
  → /bin/sh: syntax error: unexpected "(" → exit 2
```

The store had both `pnpm-lock.yaml` and `yarn.lock` committed. The build image installs from `yarn.lock` first, so pnpm was never present, while `ni` prefers `pnpm-lock.yaml`.

## How it works?

The guard now rejects anything that is not a known `ni` agent, rather than only the empty string, and reports what it got instead of failing silently.

**The yarn fallback is kept, not replaced.** Builds with `CI` set already depend on it today: in that case `ni` exits before prompting, stdout is empty, and the fallback is what makes them pass. Removing it would break working builds.

One detail worth calling out: `na` prefixes its output with `volta run` when Volta is installed, so the agent is validated from behind that prefix. Without this, every Volta user would get a warning and lose the prefix on every command.

Behaviour, verified against the real `ni` agent list:

| `na ?` output | Volta | Result | Warns |
| --- | --- | --- | --- |
| `yarn` / `pnpm` / `yarn@berry` | no | unchanged | no |
| `volta run yarn` | yes | `volta run yarn` | no |
| the confirm prompt | no | `yarn` | yes |
| the confirm prompt | yes | `volta run yarn` | yes |
| empty | no | `yarn` | no |

The last row is today's behaviour, preserved. The only intentional refinement is that the fallback now keeps the Volta prefix when Volta is present, where before it returned a bare `yarn`.

## How to test it?

- In a store with both `pnpm-lock.yaml` and `yarn.lock`, with `pnpm` not on `PATH`, run `faststore build`. Before this change it exits 2 with a shell syntax error; now it warns and builds with yarn.

Not verified locally: the `3.x` build and `jest` suite. This checkout's `node_modules` belongs to `dev`, where `@oclif/core` and the test runner differ, so `tsc -b` and `jest` fail on unrelated files. `biome check` and `tsc` are clean for the changed file, and the guard logic was exercised against the real `ni` agent list. CI covers the rest.

No test was added here: `dev` (#3422) covers the same logic with 12 cases, and an unrunnable test on this branch would be worse than none. Happy to add one if you can run it.

## References

- FAS-1199: https://vtex-dev.atlassian.net/browse/FAS-1199
- Slack thread: https://vtex.slack.com/archives/C02JMH9AYR1/p1785257117603469
- `dev` counterpart: #3422

## Checklist

**PR Title and Commit Messages**

- [x] PR title and commit messages follow the [Conventional Commits](https://www.conventionalcommits.org/en/v1.0.0/) specification

**Documentation**

- [x] PR description

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Summary by CodeRabbit

* **Bug Fixes**
  * Improved package manager detection across platforms, including Volta environments.
  * Added validation for detected package managers, with a safe fallback to Yarn when detection fails or returns unsupported, malformed, or empty output.
  * Preserved relevant Volta configuration and improved warnings for unrecognized detection results.

* **Tests**
  * Added coverage for platform-specific detection, Volta prefix handling, warnings, command failures, and fallback scenarios.

### CODE REVIEW

File: packages/cli/src/utils/commands.ts [hunk 1]
```diff
@@ -1,13 +1,46 @@
...
+
+const DEFAULT_AGENT = 'yarn'
 
 // Retrieves the package manager based on the developer lockfile, using `ni`.
 export function getPreferredPackageManager() {
-  const agent = spawnSync('na', ['?'], {
+  const resolved = spawnSync('na', ['?'], {
     encoding: 'utf8',
     shell: true,
   }).stdout.trim()
```
[the remark this answers is not in the document]
Comment (vlaux) [reply packages/cli/src/utils/commands.ts]: Split this one: took the error handling, declined the timeout.

**Taken.** `spawnSync().stdout` is `null` when the probe cannot be spawned, so `.stdout.trim()` threw a `TypeError` instead of falling back. Fixed in af951a0: it resolves to an empty string, which the existing guard already handles by returning the default agent. A failed probe is not the store's fault, so it stays silent rather than warning.

**Declined: the timeout.** The hang this guards against does not appear to be reachable. The path that could block is `ni`'s confirm prompt, and `prompts` only waits for input when stdin is a TTY. A `spawnSync` child never gets one:

```
$ node -e 'const {spawnSync}=require("node:child_process");
  console.log(spawnSync(process.execPath,["-e",
  "process.stdout.write(String(Boolean(process.stdin.isTTY)))"],{encoding:"utf8"}).stdout)'
false
```

So the prompt renders to the captured stdout and returns immediately, which is exactly how this bug produced a syntax error instead of a hang. Everything else `detect()` does is `findUp`, a `package.json` read, and a `which` lookup.

Adding a timeout would mean inventing a magic number on a maintenance branch, where a value that is too low breaks slow CI machines for no demonstrated benefit. `status` is also not worth checking separately: a non-zero exit already yields output that the widened guard rejects.

Happy to revisit if you have a concrete hang to point at.

**Review by renatomaurovtex — COMMENTED** [review]
Reviewed alongside the `dev` counterpart (#3422). This is the right shape for a `3.x` backport: minimal, additive, and it keeps the empty-output → `yarn` fallback that CI builds already depend on. I verified the branch-specific facts:

- `@antfu/ni@^0.21.12` is already a runtime dependency of `@faststore/cli` on `3.x`, and it exports both `agents` and `getVoltaPrefix` — no new dependency, no missing API.
- `chalk` and the `logger` util both exist on this branch.
- Every `3.x` caller interpolates the returned value into a `shell: true` command (`start.ts` does `spawn(`${packageManager} run serve -p ${port}`, { shell: true })`), so the one intentional refinement — the fallback keeping the Volta prefix — is safe here; there is no shell-less `spawn` on this branch to trip over it.
- The guard logic is sound: the Volta prefix is stripped *before* validating against `agents` (so Volta users don't get a spurious warning + lost prefix), the prompt garbage is truncated to one line/120 chars in the warning, and `probe.stdout?.trim() ?? ''` covers the spawn-failure case.

Two notes, neither blocking:

💬 **[pre-existing] `dependencies.ts` `=== 'npm'` comparison misses behind a Volta prefix.** With Volta, the value is `volta run npm`, so the comparison picks `add` — which happens to work since `npm add` is an alias of `npm install`. Pre-existing on this branch and properly fixed in #3422's `agent`/`command` split; no action needed here, just noting it's known.

💬 **[tests] No test on this branch.** The rationale (dev-flavored `node_modules` makes the `3.x` jest suite unrunnable locally, #3422 covers the same logic with 12 cases) is reasonable, and CI is green here — the FastStore checks ran the `3.x` suite. If a maintainer with a `3.x` checkout can drop in a small guard test before merge, even better, but I wouldn't hold the fix for it.

## Verdict: Approved

Blocking (🔴/🟠):
- None.

Non-blocking (🟡/💬):
- `npm add` alias reliance behind Volta (pre-existing, fixed properly in #3422).
- Optional guard test if someone has a runnable `3.x` checkout.

Checks to confirm before merge: `3.x` CI green (already is) · merge after or together with #3422 so the branches don't drift on the same incident.

File: packages/cli/src/utils/commands.ts [hunk 2]
```diff
@@ -1,13 +1,41 @@
...
+const DEFAULT_AGENT = 'yarn'
 
-// Retrieves the package manager based on the developer lockfile, using `ni`.
 export function getPreferredPackageManager() {
-  const agent = spawnSync('na', ['?'], {
+  const probe = spawnSync('na', ['?'], {
     encoding: 'utf8',
     shell: true,
-  }).stdout.trim()
+  })
```
[the remark this answers is not in the document]
Comment (vlaux) [reply packages/cli/src/utils/commands.ts:15]: Verdict: valid — fixed in 053b6e8f8.

Reproduced locally before fixing: in a directory with `yarn.lock` and a file named `x`, the shell expanded `na ?` into `na x`, which *ran* `yarn x` and returned its Internal Error dump as the probe output (exit 1). Without a shell, the same probe returns `volta run yarn` (exit 0).

Took the suggested shape: no shell on POSIX so `?` stays a literal argument, `shell: true` on Windows only — the `na.cmd` shim needs a shell there, and cmd.exe does not expand globs. A missing `na` now surfaces as `error` + null stdout instead of exit 127 + empty stdout; both normalise to `''` and fall back to yarn, which the existing probe-failure test asserts, and a new test pins the literal-argument call shape.

The review-body nitpick (typed, flat allowlist check via `agents.some()` + early return) landed in 7ba2ea6d9.

**Review by eduardoformiga — APPROVED** [review]
